### DISCUSSION TITLE
the error in segmented a ebsd png

### DISCUSSION THREAD

**Discussion by Yuanguo1994** [CsatiZoltan/CristalX#75] [Q&A] [answered]
Hi,
 I got an error when I made the segmentation to a ebsd microstructure. The picture is shown below:
![image](https://user-images.githubusercontent.com/79046436/124755555-6c5fd000-df23-11eb-8b3f-8b359e66f567.png)

The error message is shown like this:

segmented = GS.initial_segmentation(filtered)
__________________________________________________________________________________________________
ValueError                                Traceback (most recent call last)
<ipython-input-12-bc4a710b6d1c> in <module>
----> 1 segmented = GS.initial_segmentation(filtered)

~\CristalX\grains\segmentation.py in initial_segmentation(self, *args)
    131         else:
    132             image = self.original_image
--> 133         segment_mask = segmentation.quickshift(image)
    134         if self.__interactive_mode:
    135             io.imshow(color.label2rgb(segment_mask, self.original_image, kind='avg'))

~\Anaconda3\envs\CristalX\lib\site-packages\skimage\segmentation\_quickshift.py in quickshift(image, ratio, kernel_size, max_dist, return_tree, sigma, convert2lab, random_seed)
     61         if image.shape[2] != 3:
     62             ValueError("Only RGB images can be converted to Lab space.")
---> 63         image = rgb2lab(image)
     64 
     65     if kernel_size < 1:

~\Anaconda3\envs\CristalX\lib\site-packages\skimage\color\colorconv.py in rgb2lab(rgb, illuminant, observer)
   1090     .. [1] https://en.wikipedia.org/wiki/Standard_illuminant
   1091     """
-> 1092     return xyz2lab(rgb2xyz(rgb), illuminant, observer)
   1093 
   1094 

~\Anaconda3\envs\CristalX\lib\site-packages\skimage\color\colorconv.py in rgb2xyz(rgb)
    679     # Follow the algorithm from http://www.easyrgb.com/index.php
    680     # except we don't multiply/divide by 100 in the conversion
--> 681     arr = _prepare_colorarray(rgb).copy()
    682     mask = arr > 0.04045
    683     arr[mask] = np.power((arr[mask] + 0.055) / 1.055, 2.4)

~\Anaconda3\envs\CristalX\lib\site-packages\skimage\color\colorconv.py in _prepare_colorarray(arr, force_copy)
    148 
    149     if arr.shape[-1] != 3:
--> 150         raise ValueError("Input array must have a shape == (..., 3)), "
    151                          f"got {arr.shape}")
    152 

ValueError: Input array must have a shape == (..., 3)), got (519, 584, 4)
_________________________________________________________________________________________________________________________________

The 3D numpy.ndarray is (519, 584, 4), it is not in the shape of (...,3). Is it associated with the png I uploaded?

**Comment by CsatiZoltan** [comment, marked as answer]
Hi,

I witnessed this type of error before. This is not related to *CristalX*, rather to your input image. The image is expected to be a 3-channel image, such as RGB. So yes, the shape should be of (..., ..., 3).

**Reply by CsatiZoltan** [reply]
Most probably, the fourth dimension of your array/image consists of the same numbers. If that is the case, it encodes no useful information, so you can get rid of the fourth dimension of your array/image.

By the way, if you already have a segmented image from EBSD (as you showed above), there is no need to call the segmentation algorithms on it.

**Reply by Yuanguo1994** [reply]
Hi, CsatiZoltan

I forgot to tell you that I have solved the problem by converting the png format to bmp format. I searched in the google that png format supports 4 channel images. I am busy doing tests these days, so forgetting to respond, really sorry for that. Thanks for ur second reply. I am interested in that you said 'no need for segmentation if I have ebsd picture'. I have checked the code in the example. Do you mean that I can jump the segmentation process after I get the numpy arrays of the picture, and then go directly to the 'grain boundary found' step?

**Reply by CsatiZoltan** [reply]
> I am busy doing tests these days, so forgetting to respond, really sorry for that.

No problem, I am busy as well.

> I have solved the problem by converting the png format to bmp format.

That's another solution... I prefer PNG images due to their small size, but for such a small microstructure it doesn't matter.



What I meant was that if your image is already segmented (that's what I see in the picture you inserted at the top of the page), you already know which pixels constitute which grain. In that case, you can start turning that image into a geometry using the `cad` module of *CristalX* (similarly as you did it in #72). So you can skip the whole segmentation workflow (ending with the watershed segmentation). What remains to handle is getting rid of the thick black grain boundary pixels. What we want is that grain boundaries have "0 pixel" thickness, i.e. they are implicitly inferred from the different grain labels (like in the [example](https://github.com/CsatiZoltan/CristalX/blob/master/scripts/data/1_labelimage.png)). But those boundaries are easy to remove using dilation (I already did that for our EBSD image).

**Reply by Yuanguo1994** [reply]
> > I am busy doing tests these days, so forgetting to respond, really sorry for that.
> 
> No problem, I am busy as well.
> 
> > I have solved the problem by converting the png format to bmp format.
> 
> That's another solution... I prefer PNG images due to their small size, but for such a small microstructure it doesn't matter.
> 
> What I meant was that if your image is already segmented (that's what I see in the picture you inserted at the top of the page), you already know which pixels constitute which grain. In that case, you can start turning that image into a geometry using the `cad` module of _CristalX_ (similarly as you did it in #72). So you can skip the whole segmentation workflow (ending with the watershed segmentation). What remains to handle is getting rid of the thick black grain boundary pixels. What we want is that grain boundaries have "0 pixel" thickness, i.e. they are implicitly inferred from the different grain labels (like in the [example](https://github.com/CsatiZoltan/CristalX/blob/master/scripts/data/1_labelimage.png)). But those boundaries are easy to remove using dilation (I already did that for our EBSD image).

Thanks, I will try that!

**Reply by Yuanguo1994** [reply]
> > I am busy doing tests these days, so forgetting to respond, really sorry for that.
> 
> No problem, I am busy as well.
> 
> > I have solved the problem by converting the png format to bmp format.
> 
> That's another solution... I prefer PNG images due to their small size, but for such a small microstructure it doesn't matter.
> 
> What I meant was that if your image is already segmented (that's what I see in the picture you inserted at the top of the page), you already know which pixels constitute which grain. In that case, you can start turning that image into a geometry using the `cad` module of _CristalX_ (similarly as you did it in #72). So you can skip the whole segmentation workflow (ending with the watershed segmentation). What remains to handle is getting rid of the thick black grain boundary pixels. What we want is that grain boundaries have "0 pixel" thickness, i.e. they are implicitly inferred from the different grain labels (like in the [example](https://github.com/CsatiZoltan/CristalX/blob/master/scripts/data/1_labelimage.png)). But those boundaries are easy to remove using dilation (I already did that for our EBSD image).

Hi, Zoltan

I am looking through the CAD modulus in the documentation. However, it seems there is a lack of classes information in the chapter, seeing the picture below. Is there anywhere else I can obtain the function classes in the modulus?
![image](https://user-images.githubusercontent.com/79046436/125055581-4101f000-e09f-11eb-92d2-d4658524d7ef.png)

**Reply by CsatiZoltan** [reply]
Hi,

First of all, I recommend you to browse the [HTML version](https://cristalx.readthedocs.io/en/latest/api.html#cad) of the documentation. Although it is the same as the PDF version, some [interactive stuffs](https://cristalx.readthedocs.io/en/stable/user_documentation.html#using-the-html-documentation) are available only in the HTML output.

You are right that the API documentation of the CAD module is currently not available for the development branch of *CristalX*. I will investigate its cause. Nevertheless, you can browse the state of the latest stable version [here](https://cristalx.readthedocs.io/en/stable/api.html#module-grains.cad) (PDF output [here](https://cristalx.readthedocs.io/_/downloads/en/stable/pdf/), if you prefer that).

**Reply by Yuanguo1994** [reply]
> > I am busy doing tests these days, so forgetting to respond, really sorry for that.
> 
> No problem, I am busy as well.
> 
> > I have solved the problem by converting the png format to bmp format.
> 
> That's another solution... I prefer PNG images due to their small size, but for such a small microstructure it doesn't matter.
> 
> What I meant was that if your image is already segmented (that's what I see in the picture you inserted at the top of the page), you already know which pixels constitute which grain. In that case, you can start turning that image into a geometry using the `cad` module of _CristalX_ (similarly as you did it in #72). So you can skip the whole segmentation workflow (ending with the watershed segmentation). What remains to handle is getting rid of the thick black grain boundary pixels. What we want is that grain boundaries have "0 pixel" thickness, i.e. they are implicitly inferred from the different grain labels (like in the [example](https://github.com/CsatiZoltan/CristalX/blob/master/scripts/data/1_labelimage.png)). But those boundaries are easy to remove using dilation (I already did that for our EBSD image).

Hi, Zoltan

I still have a problem in getting the geometry of the ebsd image. As you mentioned above, if I have ebsd images, the segmentation process could be jumped. But the question is that the cad module could handle the 'labeled' image, but the ebsd image is not 'labeled' and that's my problem. I am not familiar with the image process, so don't know how to convert the rgb image to a labeled image with 2D array. Could you please give me some suggestions. I have searched the solution in google, like learning using the numpy package, but it did not solve my problem.

**Reply by CsatiZoltan** [reply]
Hi,

> the ebsd image is not 'labeled'

I have already met the very same issue before. To turn the true color (e.g. RGB) image into a labeled/indexed image, you need to determine the mapping between the unique true color values and a set of natural numbers (indices of the labeled image). E.g. if one of your grains is mostly red, with normalized RGB values of [1.0, 0.1, 0.07], you may associate the label 1 to this color. So the idea is to first determine the unique colors in your true color image, and then perform the mapping detailed above. This functionality is available in the `analysis` module of *CristalX*. Description [here](https://cristalx.readthedocs.io/en/latest/api.html#grains.analysis.truecolor2label), code [here](https://cristalx.readthedocs.io/en/latest/_modules/grains/analysis.html#truecolor2label).

**Reply by Yuanguo1994** [reply]
> Hi,
> 
> > the ebsd image is not 'labeled'
> 
> I have already met the very same issue before. To turn the true color (e.g. RGB) image into a labeled/indexed image, you need to determine the mapping between the unique true color values and a set of natural numbers (indices of the labeled image). E.g. if one of your grains is mostly red, with normalized RGB values of [1.0, 0.1, 0.07], you may associate the label 1 to this color. So the idea is to first determine the unique colors in your true color image, and then perform the mapping detailed above. This functionality is available in the `analysis` module of _CristalX_. Description [here](https://cristalx.readthedocs.io/en/latest/api.html#grains.analysis.truecolor2label), code [here](https://cristalx.readthedocs.io/en/latest/_modules/grains/analysis.html#truecolor2label).

Hi, thanks for your reply. Indeed, I have got the 2d arrays following your suggestion. And the numpy arrays were successfully handled by the polygen class in cad module. However, error occurred when I use the splinegen class. I attacded the code and the error message below.

code
_____________________________________________________
from grains.analysis import truecolor2label
data = truecolor2label(im)
data

from grains.analysis import show_label_image
data1 = show_label_image(data)
![image](https://user-images.githubusercontent.com/79046436/125940441-718512c1-3626-4c30-baec-efda62382668.png)

from grains.analysis import thicken_skeleton
data2 = thicken_skeleton(data,1)
data2
show_label_image(data2)
![image](https://user-images.githubusercontent.com/79046436/125940479-1da18add-9cc3-4c5e-9c8e-4c128375e024.png)

from grains.cad import polygonize, search_neighbor, splinegonize
polygons = polygonize(data2, search_neighbor(2, np.inf), connectivity=1)
from grains.cad import plot_polygons
plot_polygons(list(polygons.values()));
scale_figure(150)
![image](https://user-images.githubusercontent.com/79046436/125940535-d2f79b28-ee7a-4fb9-9699-88c59a18f780.png)

splinegons, _ = splinegonize(data2, search_neighbor(2, np.inf), connectivity=1, degree_min=3, degree_max=3, continuity='C0', tol=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-54-a7bcf3f91edb> in <module>
----> 1 splinegons, _ = splinegonize(data2, search_neighbor(2, np.inf), connectivity=1, degree_min=3, degree_max=3, continuity='C0', tol=1)

~\CristalX\grains\cad.py in splinegonize(label_image, neighbor_search_algorithm, connectivity, detect_boundaries, degree_min, degree_max, continuity, tol)
    722         order, _, _, = branches2boundary(index_list(branch_coordinates, branches))
    723         # Construct the surface and the boundary of the region
--> 724         splinegon, boundary = region_as_splinegon(index_list(boundary_splines, order))
    725         # Save the created regions and their boundaries
    726         splinegons[region] = splinegon

~\CristalX\grains\cad.py in region_as_splinegon(boundary_splines)
    604     # The planar surface (face) is stretched by the wire
    605     splinegon = BRepBuilderAPI_MakeFace(boundary.Wire())
--> 606     _face_error(splinegon.Error())
    607     return splinegon.Face(), boundary.Wire()
    608 

~\CristalX\grains\cad.py in _face_error(error_code)
   1135         raise ValueError('Empty constructor was used. The face does not yet exist.')
   1136     elif error_code == 2:
-> 1137         raise ValueError('The face cannot be constructed from a non-planar wire.')
   1138     elif error_code == 3:
   1139         raise ValueError('Curve projection failed.')

ValueError: The face cannot be constructed from a non-planar wire

**Reply by CsatiZoltan** [reply]
For some reason, the created wire is not planar. This is unexpected because my whole code works in 2D exclusively. I will have to look at it in the debugger to find out what causes this error. Please send me your labeled image, and I will check it on my side once I have time.

**Reply by Yuanguo1994** [reply]
> For some reason, the created wire is not planar. This is unexpected because my whole code works in 2D exclusively. I will have to look at it in the debugger to find out what causes this error. Please send me your labeled image, and I will check it on my side once I have time.

Hi,  sorry for the delaying reply. I was doing some tests this afternoon.
I have solved the problem, it is because of my picture, not the code. 
There are some white edges of the picture causing the problem. After I cropped the picture, the problem was solved. Another trouble is that there are two grains not identified for the microstructure. Becasue last time I commented out the line for making the hole, so I cant repair the grains in Salome. Is there any other way to solve the problem?
I attactched the original image with white edges and the microstructure with two grains not identified below:
![image](https://user-images.githubusercontent.com/79046436/125989279-32e0ee70-1c1c-479d-b801-ec5b06cdfb7a.png)
![image](https://user-images.githubusercontent.com/79046436/125989309-b2fafa83-41b4-413f-beba-4962b3477b82.png)

**Reply by CsatiZoltan** [reply]
> Another trouble is that there are two grains not identified for the microstructure.

That is not unexpected. The algorithm described in Section 2.3 (especially step 2) of [our paper](https://www.sciencedirect.com/science/article/pii/S2352711021000145) explains the heuristics. What you can do is experiment with the parameters, whose meanings are explained in details in the section referenced above, and the function signatures are described in the [documentation](https://cristalx.readthedocs.io/en/stable/functions/grains.cad.skeleton2regions.html#grains.cad.skeleton2regions) (check the links there as well).
You should modify the parameters so as to achieve the best result you can, when compared with the segmented image. Due to the heuristic nature of the algorithm, you may still end up with grains not identified. The goal is to have as few unidentified grains as possible to minimize the manual work in Salome. If you only have two unidentified grains, it is a matter of minutes to create them manually in Salome.



> Becasue last time I commented out the line for making the hole, so I cant repair the grains in Salome.

The second clause does not follow from the first one.
1. The effect of commenting out the line that calls the function that makes the holes is that you will have overlapping grains for the non-simply connected domains. You can eliminate them manually in Salome or you can wait for some time until I fix that in the code (by updating to a newer version of PythonOCC).
2. You can always repair grains in Salome by defining a closed wire from the sides of a missing/unidentified grain, and then defining a face surrounded by that wire. That face will be your grain you want. You can follow the process according to Point 4 in the [guide](https://cristalx.readthedocs.io/en/stable/salome.html#geometry).

**Reply by Yuanguo1994** [reply]
Hi, I have followed the process to repair the geometry, but still I cant finish the repair work. Let me take your example microstructure for instance.
The hole in the microstructure is shown below (the first picture), as writen in the documentation, I selected the sides edges of the hole to create the closed contour (second picture), but more holes appeared, shown in the third picture. And then I suppressed the hole, the hole I want to supress was not sucessfully removed, shown in the fourth picture. 
 
![image](https://user-images.githubusercontent.com/79046436/126071298-4a731f7b-39da-4152-960a-1068a2fae33f.png)
![image](https://user-images.githubusercontent.com/79046436/126071531-56f95485-2c3c-4f13-8f4f-5874e5907a10.png)
![image](https://user-images.githubusercontent.com/79046436/126071550-fb36e513-9e4a-4ade-8674-e67e43ee86ff.png)
![image](https://user-images.githubusercontent.com/79046436/126071566-1109921c-efe8-4775-b137-b53881aeb258.png)

**Reply by CsatiZoltan** [reply]
> The hole in the microstructure is shown below (the first picture), as writen in the documentation, I selected the sides edges of the hole to create the closed contour (second picture)

I should have emphasized in the documentation that closing a hole makes sense when you have a grain that contains a smaller grain. In that case, you can select the larger grain, which contains the hole, and suppress that hole. In your case, you can see that the hole (corresponding to an unidentified grain) is not an embedded grain (it has more than one neighbor). Hence, there is no hole to suppress. What you rather want is creating a *new* grain in place of that hole. You can do it in a few seconds by creating a closed wire and then a planar face on it. I demonstrate it on a different microstructure, but you can see the idea. Check the GIF below.
![Make_grain](https://user-images.githubusercontent.com/12742890/126076717-47a07f58-0f5d-4f4b-821f-b9635aad0d7e.gif)

> more holes appeared

In most cases, that's just a visualization glitch -- something that happens with other software too (e.g. CAD Assistant). If the subsequent mesh does not contain such holes, you are good to go. If, in some rare cases, those are actual holes, you will need to repair the geometry by sewing it. But for now, don't bother about it.

**Reply by Yuanguo1994** [reply]
> > The hole in the microstructure is shown below (the first picture), as writen in the documentation, I selected the sides edges of the hole to create the closed contour (second picture)
> 
> I should have emphasized in the documentation that closing a hole makes sense when you have a grain that contains a smaller grain. In that case, you can select the larger grain, which contains the hole, and suppress that hole. In your case, you can see that the hole (corresponding to an unidentified grain) is not an embedded grain (it has more than one neighbor). Hence, there is no hole to suppress. What you rather want is creating a _new_ grain in place of that hole. You can do it in a few seconds by creating a closed wire and then a planar face on it. I demonstrate it on a different microstructure, but you can see the idea. Check the GIF below.
> ![Make_grain](https://user-images.githubusercontent.com/12742890/126076717-47a07f58-0f5d-4f4b-821f-b9635aad0d7e.gif)
> 
> > more holes appeared
> 
> In most cases, that's just a visualization glitch -- something that happens with other software too (e.g. CAD Assistant). If the subsequent mesh does not contain such holes, you are good to go. If, in some rare cases, those are actual holes, you will need to repair the geometry by sewing it. But for now, don't bother about it.

Hi, 

Thanks for the reply, after modifying the parameters of search_neighbor function, I have got a good geometry, shown in the first picture. However, after I did explode to create the faces, some gaps between grains appeared (the second and third picture), even though I tried different limit tolerance. I dont know how to solve the problem, could u please give me some advice?
![image](https://user-images.githubusercontent.com/79046436/126493963-e0b7b450-26f4-428c-a272-dd98bcac4d39.png)
![image](https://user-images.githubusercontent.com/79046436/126494118-b9be5eff-f20b-4825-9cca-8fdab33592cb.png)
![image](https://user-images.githubusercontent.com/79046436/126494175-75f5ae9b-39ce-48b5-8435-9c0b9e081475.png)

**Reply by CsatiZoltan** [reply]
Hi,

True gaps should not appear. I have not noticed anything like this done by the explode tool in Salome. My suspicion is still that this is a visualization glitch. If you are worried about it, export the exploded geometry (select it in the *Object Browser* and then *File* --> *Export* --> *STEP*) and open it in another CAD program/viewer, such as [FreeCAD](https://www.freecadweb.org/) or [CAD Assistant](https://www.opencascade.com/products/cad-assistant/). However, if you want to generate a mesh on the microstructure, you can simply go to the *Mesh* module in Salome and create the mesh. If there was a real gap between the grains, the mesh will not be conforming -- an easy way to check whether your gap is real or not.

If you have further questions on meshing or other topics not related to this thread, please open a new one.

**Reply by Yuanguo1994** [reply]
Yes, the geometry could be successfully meshed, so it should be visualization glitch. Thanks for ur so many patient answers. I might have some questions about export the mesh to the format reqiured by FEA software such  as abaqus and fepx. I will open a new discussion if I need ur help.
